launcher/src/test/java/org/apache/spark/launcher/LauncherServerSuite.java

```diff
@@ -152,6 +175,102 @@ public void testTimeout() throws Exception {
     }
   }
 
+  @Test
+  public void testSparkSubmitVmShutsDown() throws Exception {
```

Comment:
This test looks very complicated. Why not:

- create a server and a client that connects to it
- initiate the connection (send "Hello")
- close the client socket
- verify the server socket fires the expected event

This would be very similar to `testCommunication` above until about L112. After that you just close the client socket and wait for the event on the handle.